### PR TITLE
Bifurcate Dependency Lists [skip-gpuci]

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR uses the [`rapids-dependency-file-generator`](https://github.com/rapidsai/dependency-file-generator/) to handle sourcing dependencies. Similar to https://github.com/rapidsai/rmm/pull/1073/, this PR introduces a GitHub Action that enforces consistency between the new `dependencies.yaml` file and the generated conda environment for developers.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by bdice** [comment]
@ajschmidt8 I think this PR is close to ready -- for the `environment.yaml` only. I have a few TODOs left but nothing crucial. Currently the `environment.yaml` is a hybrid of cudf + dask-cudf + custreamz + cudf-kafka and we'll need some follow-up work to separate all that out once we get to generating `meta.yaml` recipes and `requirements.txt`. However, that can wait because the GitHub Actions CI probably only cares about the combined environment for now.

File: dependencies.yaml [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+          - librdkafka=1.7.0
+          - protobuf>=3.20.1,<3.21.0a0
+          - python>=3.8,<3.10
+    specific:
+      - output_types: [conda, requirements]
+        matrices:
+        - matrix:
+            cuda: "11.5"
+          packages:
+            - cuda-python>=11.7.1,<12.0
```
Comment (ajschmidt8) [inline dependencies.yaml:51]: move this to `common`

File: dependencies.yaml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      - output_types: conda
+        matrices:
+        - matrix:
+            arch: x86_64
+          packages:
+            - gcc_linux-64=9.*
+            - sysroot_linux-64==2.17
+        - matrix:
+            arch: aarch64
```
Comment (ajschmidt8) [inline dependencies.yaml:60]: ToDo: check `arch` string

File: dependencies.yaml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+          packages:
+            - gcc_linux-aarch64=9.*
+            - sysroot_linux-aarch64==2.17
+      - output_types: conda
+        matrices:
+        - matrix:
+            cuda: "11.5"
+          packages:
+            - cudatoolkit=11.5
```
Comment (ajschmidt8) [inline dependencies.yaml:69]: Break `cudatoolkit` package out into its own `dependencies` key.

Remove `11.0` from `rmm`'s `dependencies.yaml`.

File: dependencies.yaml [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      - output_types: conda
+        packages:
+          - cubinlinker
+          - cupy>=9.5.0,<12.0.0a0
+          - pip
+          - pip:
+              - git+https://github.com/python-streamz/streamz.git@master
+              - pyorc
+          - ptxcompiler
+          - rmm=22.12.*
```
Comment (ajschmidt8) [inline dependencies.yaml:107]: update `update-version.sh`.

File: .github/workflows/dependency-files.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+name: dependency-files
```
Comment (ajschmidt8) [inline .github/workflows/dependency-files.yml:1]:
```suggestion
name: pr
```

File: .github/workflows/dependency-files.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+on:
+  pull_request:
+
+jobs:
+  check-generated-files:
+    uses: rapidsai/shared-action-workflows/.github/workflows/dependency-files.yaml@main
```
Comment (ajschmidt8) [inline .github/workflows/dependency-files.yml:8]:
```suggestion
  checks:
    secrets: inherit
    uses: rapidsai/shared-action-workflows/.github/workflows/checks.yaml@main
    with:
      enable_check_size: false
      enable_check_size: false
```

File: dependencies.yaml (hunk 2 above)
Comment (bdice) [reply dependencies.yaml:55]: Checked. The string is `aarch64` as expected.

Comment (ajschmidt8) [reply dependencies.yaml:55]: Confirmed [here](https://github.com/rapidsai/rmm/actions/runs/3395618780/jobs/5681030561#step:6:94)

File: dependencies.yaml [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      - notebook
+  test:
+    output: none
+    includes:
+      - cudatoolkit
+      - test
+channels:
+  - rapidsai
+  - conda-forge
+# Not sure whether we need these:
```
Comment (bdice) [inline dependencies.yaml:24]: @galipremsagar Can I get your insight on this? Do we (always/sometimes/never) use the `dask/label/dev` channel?

Comment (galipremsagar) [reply dependencies.yaml:24]: `dask/label/dev` is only needed during development through code-freeze. And `dask/label/dev` has to be dropped just like we drop `rapidsai-nightly` during release runs.

Comment (bdice) [reply dependencies.yaml:24]: @ajschmidt8 Same question for `rapidsai-nightly` -- when (if ever) do we need `rapidsai-nightly`?

I don't think we need the `nvidia` channel for cudf, but I think we will need it for cuML and cuGraph.

**Review by ajschmidt8 — APPROVED** [review]

File: dependencies.yaml [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            packages:
+              - cupy-cuda11x -f https://pip.cupy.dev/aarch64  # TODO: Verify that this works.
+  test_python:
+    common:
+      - output_types: [conda, requirements]
+        packages:
+          - aiobotocore>=2.2.0
+          - boto3>=1.21.21
+          - botocore>=1.24.21
+          - dask-cuda=22.12.*  # TODO: Verify this is a testing (not run) dependency.
```
Comment (bdice) [inline dependencies.yaml:193]:
```suggestion
          - dask-cuda=22.12.*
```

File: conda/environments/all_cuda-115_arch-x86_64.yaml [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# This file is generated by `rapids-dependency-file-generator`.
+# To make changes, edit ../../dependencies.yaml and run `rapids-dependency-file-generator`.
+channels:
+- rapidsai
+- conda-forge
```
Comment (bdice) [inline conda/environments/all_cuda-115_arch-x86_64.yaml:5]:
```suggestion
- rapidsai-nightly
- dask/label/dev
- conda-forge
- nvidia
```

**Review by ajschmidt8 — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

File: dependencies.yaml [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    includes:
+      - build
+      - develop
+      - py_version
+channels:
+  - rapidsai
+  - rapidsai-nightly
+  - dask/label/dev
+  - conda-forge
+  - nvidia
```
Comment (ajschmidt8) [inline dependencies.yaml:37]:
```suggestion
  - conda-forge
  - dask/label/dev
  - nvidia
  - rapidsai
  - rapidsai-nightly
```

**Pull request merged**